File: backend/src/main/java/com/sss/tally/domain/payment/entity/Payment.java
package com.sss.tally.domain.payment.entity;

import java.time.LocalDateTime;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import com.sss.tally.api.payment.dto.PaymentDto;
import com.sss.tally.domain.category.entity.Category;
import com.sss.tally.domain.member.entity.Member;
import com.sss.tally.domain.paymentunit.entity.PaymentUnit;
import com.sss.tally.domain.travel.entity.Travel;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@Builder
@EntityListeners(AuditingEntityListener.class)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class Payment {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long paymentId;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "member_id")
	private Member memberId;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "travel_id")
	private Travel travelId;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "category_id")
	private Category categoryId;

	@Column(nullable = false)
	private int amount;

	@Column(nullable = false)
	private Double ratio;

	@Column(nullable = false)
	private String paymentUuid;

	@Column(nullable = false)
	private LocalDateTime paymentLocalDate;

	@CreatedDate
	@Column(nullable = false)
	private LocalDateTime paymentKoreaDate;

	@Column(nullable = false)
	private String paymentMemo;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private PaymentMethodEnum paymentMethod;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "payment_unit_id")
	private PaymentUnit paymentUnitId;

	@Column(nullable = false)
	private Boolean visible;

	@Column(nullable = false)
	private Boolean status;

	@Column(nullable = false)
	private String paymentName;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private CalculateStatusEnum calculateStatus;


	public static Payment of(PaymentDto.PaymentManualDto payPaymentManualDto, Member member, Travel travel, Category category, PaymentUnit paymentUnit, String uuid, PaymentMethodEnum paymentMethod, LocalDateTime dateTime){
		return Payment.builder()
			.memberId(member)
			.travelId(travel)
			.paymentUnitId(paymentUnit)
			.categoryId(category)
			.paymentUuid(uuid)
			.amount(payPaymentManualDto.getAmount())
			.paymentLocalDate(dateTime)
			.paymentKoreaDate(dateTime)
			.paymentMemo(payPaymentManualDto.getMemo())
			.paymentMethod(paymentMethod)
			.ratio(payPaymentManualDto.getRatio())
			.visible(payPaymentManualDto.isVisible())
			.paymentName(payPaymentManualDto.getTitle())
			.calculateStatus(CalculateStatusEnum.NONE)
			.status(false)
			.build();
	}

	public void updateCalculateStatusEnum(CalculateStatusEnum calculateStatus) {
		this.calculateStatus = calculateStatus;
	}

	public void updateMemo(String memo){
		this.paymentMemo = memo;
	}

	public void updatePayment(int amount, Double ratio, PaymentUnit paymentUnitId, Category category, LocalDateTime paymentDateTime, String memo, boolean visible, String title){
		this.paymentName = title;
		this.amount = amount;
		this.ratio = ratio;
		this.categoryId = category;
		this.paymentUnitId = paymentUnitId;
		this.paymentLocalDate = paymentDateTime;
		this.paymentKoreaDate = paymentDateTime;
		this.paymentMemo = memo;
		this.visible = visible;
	}

	public void updatePaymentAuto(PaymentDto.PaymentCardUpdateDto paymentCardUpdateDto, Category category){
		this.categoryId = category;
		this.paymentMemo = paymentCardUpdateDto.getMemo();
		this.visible = paymentCardUpdateDto.isVisible();
	}

	public static Payment of(PaymentDto.PaymentListRespDto paymentListRespDto, Member member, Travel travel, Category category, PaymentUnit paymentUnit, LocalDateTime datetime){
		return Payment.builder()
			.paymentName(paymentListRespDto.getContent())
			.amount(paymentListRespDto.getAmount())
			.memberId(member)
			.travelId(travel)
			.ratio(1.0)
			.categoryId(category)
			.paymentUuid(paymentListRespDto.getTransferUuid())
			.paymentMethod(PaymentMethodEnum.CARD)
			.paymentUnitId(paymentUnit)
			.visible(true)
			.status(false)
			.paymentMemo("")
			.paymentKoreaDate(datetime)
			.paymentLocalDate(datetime)
			.calculateStatus(CalculateStatusEnum.NONE)
			.build();
	}
}

File: frontend/app/model/payment.ts
export interface DirectPayReq {
	amount: number; // 결제 금액(Long)
	category: number; // category type id(Long)
	memo: string; // 한줄 메모 (String)
	payment_date_time: string; // 결제 시간(yyyy-MM-dd HH:mm) String!!!
	payment_participants: Array<DirectPayMember>;
	payment_type: string; // 결제 타입 "cash" or "card" String
	payment_unit_id: number; // 결제 단위 (달러, 엔 등) Long
	ratio: number; // 비율 amount에 원화가 되기 위해 곱해야하는 값.
	title: string; // 거래처 String
	travel_id: number; // 여행 id (Long)
	visible: boolean;
}
export interface Payment {
	payment_uuid: string; // 결제 uuid (string)
	category_id: number; // 카테고리 id(Long)
	amount: number; // 결제 총 금액(int)
	payer: string; //결제자 uuid
	payment_date: string; // 결제 시간(String)
	payment_memo: string; // 결제 상세 메모(String)
	payment_method: string; // 결제 방식(String)
	payment_unit: string; // 결제 단위(String)
	visible: boolean; // 공개여부(boolean)
	payment_name: string; // 예금주(String)
	calculate_status: string;
	participants: string[];
}

export interface DirectPayMember {
	amount: number;
	member_uuid: string;
}
// 결제 추가 시 사용하는 멤버
export interface SelectPayMember {
	amount: number;
	member_uuid: string;
	checked: boolean;
	member_nickname: string;
	image: string;
}

//결제 수정시 사용되는 멤버
export interface ModMember {
	member_uuid: string;
	amount: number;
	nickname: string;
	profile_image: string;
	payer: boolean;
	with: boolean;
}

//결제 내역  Response
export interface PaymentDetailRes {
	payment_uuid: string; // 결제 uuid(String)
	category: number; // 카테고리 id(Long)
	memo: string; // 메모(String)
	visible: boolean; // 공개여부(boolean)
	amount: number; // 총금액(int)
	payment_unit: string; // 결제 단위(String)
	payment_date: string; // 결제 시간(String)
	payment_participants: ModMember[];
}

export interface PaymentModifyReq {
	amount?: string;
	category: number;
	memo: string;
	payment_date_time?: string;
	payment_participants: DirectPayMember[];
	payment_uuid: string;
	travel_id: number;
	title?: string;
	visible: boolean;
	ratio?: number;
	payment_unit_id?: number;
}
